### ISSUE TITLE
OpenGL Issue (Raspberry pi4)

### ISSUE THREAD

**Issue by freakout242** [Genymobile/scrcpy#4113]
**Environment**
 - OS: Debian (Raspberry pi 4)
 - scrcpy version: [1.17
 - installation method: [manual build via installation instructions]
 - device model: OnePlus HD1907
 - Android version: 12

**Describe the bug**
Error when opening scrcpy.  Says it needs a better version of OpenGL. My question is, how do I update my OpenGL drivers on my phone?

EDIT: It's not my phone's OpenGL it's the Raspberry Pi. Question still applies, how can I update the OpenGL from 2.1 to 3.0 ? Thanks

```
INFO: Renderer: opengl
INFO: OpenGL version: 2.1 Mesa 20.3.5
WARN: Trilinear filtering disabled (OpenGL 3.0+ or ES 2.0+ required)
INFO: Initial texture: 1080x2400
WARN: Device disconnected
WARN: Killing the server...

```

**Comment by rom1v** [comment]
> Error when opening scrcpy. Says it needs a better version of OpenGL

You can ignore this line, it's just that you won't get trilinear filtering.

The real problem is:

> WARN: Device disconnected

https://github.com/Genymobile/scrcpy/blob/master/FAQ.md#device-disconnected

On a Raspberry pi it might be due to insufficient power supply.

**Comment by freakout242** [comment]
Thanks for the reply. My power supply is more than adequate to handle one USB device. (I learned the hard way not to use cheap USB chargers) When I connect my phone, I can browse my phone's file system over the RPI. 

It is defiantly connected. The USB cable is also a known good cable.

here is everything from when I run scrcpy.

`pi@raspberrypi:~ $ scrcpy
INFO: scrcpy 1.17 <https://github.com/Genymobile/scrcpy>
/usr/share/scrcpy/scrcpy-server: 1 fil...shed. 2.0 MB/s (34015 bytes in 0.017s)
[server] INFO: Device: OnePlus HD1907 (Android 12)
[server] ERROR: Exception on thread Thread[main,5,main]
java.lang.AssertionError: java.lang.reflect.InvocationTargetException
	at com.genymobile.scrcpy.wrappers.SurfaceControl.setDisplaySurface(SurfaceControl.java:75)
	at com.genymobile.scrcpy.ScreenEncoder.setDisplaySurface(ScreenEncoder.java:243)
	at com.genymobile.scrcpy.ScreenEncoder.internalStreamScreen(ScreenEncoder.java:91)
	at com.genymobile.scrcpy.ScreenEncoder.streamScreen(ScreenEncoder.java:60)
	at com.genymobile.scrcpy.Server.scrcpy(Server.java:80)
	at com.genymobile.scrcpy.Server.main(Server.java:252)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:391)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.genymobile.scrcpy.wrappers.SurfaceControl.setDisplaySurface(SurfaceControl.java:73)
	... 7 more
Caused by: java.lang.IllegalArgumentException: displayToken must not be null
	at android.view.SurfaceControl$Transaction.setDisplaySurface(SurfaceControl.java:3302)
	at android.view.SurfaceControl.setDisplaySurface(SurfaceControl.java:2278)
	... 9 more
INFO: Renderer: opengl
INFO: OpenGL version: 2.1 Mesa 20.3.5
WARN: Trilinear filtering disabled (OpenGL 3.0+ or ES 2.0+ required)
INFO: Initial texture: 1080x2400
WARN: Device disconnected
WARN: Killing the server...`

**Comment by rom1v** [comment]
> scrcpy 1.17

Android 12 requires at least scrcpy 1.18. You should install the latest version (currently 2.1).

**Comment by freakout242** [comment]
> > scrcpy 1.17
> 
> Android 12 requires at least scrcpy 1.18. You should install the latest version (currently 2.1).

I ran the git pull and ./install_release.sh. Says I'm up to date. How can I upgrade to 2.1?

**Comment by rom1v** [comment]
You must uninstall the one from your distribution:

```
apt remove scrcpy
```

**Comment by freakout242** [comment]
Fixed: After downloading 2.1, I had no issue connecting. Thanks for the help.

**Comment by rahe13** [comment]
Hallo Leute,
kann mir jemand sagen wo mein Problem ist, ich sitze schon die ganze Nacht daran.

 ra@ra:~ $ cd scrcpy
ra@ra:~/scrcpy $ scrcpy
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     --> (tcpip)  192.168.2.195:5555              device  SM_G960F
/usr/local/share/scrcpy/scrcpy-server: 1 file pushed. 2.0 MB/s (90224 bytes in 0.043s)
[server] INFO: Device: [samsung] samsung SM-G960F (Android 10)
[server] WARN: Audio disabled: it is not supported before Android 11
INFO: Renderer: opengl
INFO: OpenGL version: 1.4 Mesa 20.3.5
WARN: Trilinear filtering disabled (OpenGL 3.0+ or ES 2.0+ required)
WARN: Demuxer 'audio': stream explicitly disabled by the device
